### PR TITLE
Use static metrics for storage/engine

### PR DESCRIPTION
**Pull request by breeswish**
## What have you changed? (mandatory)

Make metrics in `storage/engine/metrics.rs` static, i.e. avoid `with_label_values`.

Previously we have already did this for `server.rs`, with a 5%~7% performance improvement.

This PR utilizes the [latest feature](https://github.com/pingcap/rust-prometheus/pull/182) of static metric macro provided by rust-prometheus, which is, ensure type-safety when label values of static metrics is unknown at compile-time.

## What are the type of the changes? (mandatory)

- Improvement (non-breaking change which is an improvement to an existing feature)

## How has this PR been tested? (mandatory)

No test is needed. Once it is compiled successfully, it works.

## Does this PR affect documentation (docs/docs-cn) update? (mandatory)

No.

## Does this PR affect tidb-ansible update? (mandatory)

No. However notice that some tag name has been changed by adding a `err_` prefix.

## Refer to a related PR or issue link (optional)

This PR contains changes from https://github.com/pingcap/tikv/pull/3226 , so you may want to merge that PR first.

### CODE REVIEW

**Comment by breeswish** [comment]
@overvenus @siddontang PTAL

File: src/storage/mod.rs [hunk 1]
```diff
@@ -1397,28 +1397,60 @@ quick_error! {
...
+    KeyNotInRegion,
+    StaleEpoch,
+    ServerIsBusy,
+    StaleCommand,
+    StoreNotMatch,
+    RaftEntryTooLarge,
+    Other,
+}
+
+impl Display for ErrorHeaderKind {
```
Comment (breeswish) [inline src/storage/mod.rs]: note: this `Display` impl is supposed to be removed in future, once coprocessor is using static metrics as well.

Comment (ngaut) [reply src/storage/mod.rs]: Please add a todo.

File: src/storage/engine/metrics.rs [hunk 2]
```diff
@@ -11,18 +11,80 @@
...
 
-use prometheus::{exponential_buckets, HistogramVec, IntCounterVec};
+use prometheus::exponential_buckets;
+use prometheus_static_metric::*;
+
+use storage::ErrorHeaderKind;
+
+make_static_metric! {
+    pub label_enum RequestStatusKind {
+        all,
```
Comment (Hoverbear) [inline src/storage/engine/metrics.rs]: I am surprised clippy does not complain about these being lowercase.

**Comment by huachaohuang** [comment]
LGTM

File: src/storage/engine/metrics.rs (hunk 2 above)
Comment (breeswish) [reply src/storage/engine/metrics.rs]: Ha. That's because the macro will add a `#[allow]` attribute.

**Pull request merged**